udf-compiler/src/main/scala/com/nvidia/spark/udf/LambdaReflection.scala

```diff
@@ -152,7 +151,9 @@ object LambdaReflection {
   }
 
   def getClass(name: String): Class[_] = {
-    ShimLoader.loadClass(name)
+    // scalastyle:off classforname
+    Class.forName(name, true, Thread.currentThread().getContextClassLoader)
```

Comment:
Sorry by this do you mean those classes defined in our jar? If so, I’m fairly certain the answer is yes. The classes we load are external to our jar.

Reply:
I am somewhat uneasy about sweeping under the rug the bug inside ShimLoader.loadClass, and introducing the divergent call path. 

On the other hand this PR fixes UDFs that don't require access to classes inside the jar. 

It seems we need to think whether we should always prioritize the thread context classloader before doing the current logic of classloader selection in ShimLoader.loadClass. So probably we can go ahead and merge this fix and file an issue for prioritizing the thread context classloader

Would also be good to hear more opinions given the release timing

Reply:
@gerashegalov what exactly is the bug with ShimLoader.loadClass? Doesn't it work with shims (as intended)? 

I guess my point is why should we use a class loader for shims, when this has nothing to do with shims?

Reply:
> we should always prioritize the thread context classloader before doing the current logic of classloader selection in ShimLoader.loadClass. 

For shims, do you have guarantees that the shimmed classes will be in the thread context classloader? Or is this adding another lookup?

Reply:
> what exactly is the bug with ShimLoader.loadClass? Doesn't it work with shims (as intended)?

ShimLoader's classloader is designed to manage classloading for all of the plugin except when we have no choice with the config processed directly by Spark. It loads most of the classes including unshimmed classes by delegating to the parent classloader. And it checks for the thread context  classloader as a [fallback](https://github.com/NVIDIA/spark-rapids/blob/branch-22.04/sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala#L158). We might have to rethink the precedence, maybe just in case of REPLs

Reply:
If I understand the code correctly, once you find a class loader you set the "shim class loader" and that gets used going forward. Does it seem we need to add a fallback there?

https://github.com/NVIDIA/spark-rapids/blob/branch-22.04/sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala#L358

I believe you are suggesting that one would try to find the class in the shim class loader, and if not we try the thread's context classloader.

I can see why something like this would be good, but I don't know enough about class loaders. It doesn't seem that shims on their own are having issues, but perhaps something like this would help other cases?

Reply:
It also turns that it would be supported from the classloading standpoint even if it was required because we end up modifying an ancestor of the REPL thread context classloader, so it can still reach classes even from shimmed areas:
```
scala> Thread.currentThread.getContextClassLoader.loadClass("org.apache.spark.sql.rapids.shims.spark321.RapidsShuffleInternalManager")
res3: Class[_] = class org.apache.spark.sql.rapids.shims.spark321.RapidsShuffleInternalManager
```

The other design point of routing everything via ShimLoader was just working around the fact that we have only one touch surface and never miss a chance to modify the relevant Spark classloader, given that there are call paths activated before the plugin initialization. This is not the case in this context. 

I suggest we add some comment above this Class.forName.
// Don't use ShimLoader.loadClass because in the REPL use case the classes
// compiled by REPL are accessible via Thread context classloader but not by the parent 
// classloader detected by the plugin.
// This code is executed after the Plugin and ShimLoader has already been initialized.  
// REPL classes are able to interact with the Plugin classes because the the REPL Thread context classloader is a descendant of the Plugin classloader.

Reply:
@gerashegalov you got it (https://github.com/NVIDIA/spark-rapids/pull/5030/commits/1020f54439900c0c76aa08245975f45340ad44c6)